Check if generator has next normal sample cache methods in normal_distribution

pbelevich opened the pull request:
This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39816 Check if generator has next normal sample cache methods in normal_distribution**

Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

pbelevich committed:
Check if generator has next normal sample cache methods in normal_distribution

[ghstack-poisoned]

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"




[ghstack-poisoned]

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng

Fixes #39618




[ghstack-poisoned]

pbelevich committed:
Update on "[WIP] Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng

Fixes #39618




[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/DistributionsHelper.h:
```diff
@@ -126,6 +126,65 @@ struct uniform_real_distribution {
     T to_;
 };
 
+#define DISTRIBUTION_HELPER_GENERATE_HAS_MEMBER(member)              \
```
This is a really big macro. Is there really no way to do this without macro?

pbelevich replied:
hm... this macro is 8 loc

ezyang replied:
Apparently, this macro is a version of an old-style detection idiom in C++. In C++17 you could have just use `std::is_detected` to do this, no macros necessary. The way we generally prefer to backport features like this is to backport an implementation of the macro in question (e.g., the reference implementation from cppreference) so that we can easily switch to using the standard library version when we eventually upgrade to C++.

pbelevich replied:
If you' re saying that I have to implement `std::is_detected` then can I do it in follow up PR and leave this as is?

ezyang commented:
@smessmer would you mind helping review the C++ portions of this?

ezyang commented on aten/src/ATen/core/DistributionsHelper.h:
```diff
@@ -144,41 +203,29 @@ struct normal_distribution {
+    if (std::is_same<T, double>::value && maybe_get_next_double_normal_sample(generator, &ret)) {
+      ret = transformation::normal(ret, mean, stdv);
+      return ret;
+    } else if (maybe_get_next_float_normal_sample(generator, &ret)) {
```
This looks wrong. T is double I should never try to get the next float sample, right?

pbelevich replied:
addressed

ezyang commented:
This all seems a bit finely balanced for my taste. In particular, if I misspell the method name for getting/setting the cache values, I will just silently start wasting random numbers with no indication from the compiler at all.

You have control over all of the classes in question. Can't you do a more direct type test, e.g., forward declare some generator variants and just test if the type is one of those generator types before using the method names?

smessmer commented on aten/src/ATen/core/DistributionsHelper.h:
```diff
@@ -126,6 +126,65 @@ struct uniform_real_distribution {
 };
 
+#define DISTRIBUTION_HELPER_GENERATE_HAS_MEMBER(member)              \
+template <typename T>                                                \
```
This can be more easily implemented with `guts::void_t`, something like

```cpp
template<class T, class Enable = void> struct has_member_X : std::false_type {};
template<class T> struct has_member_X<T, guts::void_t<T::X>> : std::true_type {};
```

smessmer commented on aten/src/ATen/core/DistributionsHelper.h:
```diff
@@ -126,6 +126,65 @@ struct uniform_real_distribution {
+#define DISTRIBUTION_HELPER_GENERATE_NEXT_NORMAL_METHODS(TYPE)                                      \
+                                                                                                    \
+template <typename RNG, typename ret_type,                                                          \
+          typename std::enable_if<(                                                                 \
```
nit: use `enable_if_t` and you don't need the `typename` or the `::type`

pbelevich replied:
addressed

smessmer commented on aten/src/ATen/core/DistributionsHelper.h:
```diff
@@ -126,6 +126,65 @@ struct uniform_real_distribution {
+  return false;                                                                                     \
+}                                                                                                   \
+                                                                                                    \
+template <typename RNG, typename ret_type,                                                          \
```
Instead of SFINAE, you could use `guts::if_constexpr` here to combine both into one function and reduce the amount of code. same for the setter function below.

pbelevich replied:
not possible for now, because of "RuntimeError: CUDA error: unspecified launch failure"

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]

pbelevich commented:
> This all seems a bit finely balanced for my taste. In particular, if I misspell the method name for getting/setting the cache values, I will just silently start wasting random numbers with no indication from the compiler at all.
> 
> You have control over all of the classes in question. Can't you do a more direct type test, e.g., forward declare some generator variants and just test if the type is one of those generator types before using the method names?

@ezyang in future - yes, this PR is just to unblock CSPRNG release

smessmer reviewed (approved):
Looks good to me now. Please make sure @ezyang's concerns are addressed before landing.

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]

ezyang commented:
> in future - yes, this PR is just to unblock CSPRNG release

OK, in that case, write into the code what the intended refactor is so that we don't forget about it.

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]

pbelevich committed:
Update on "Check if generator has next normal sample cache methods in normal_distribution"

This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618


Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

[ghstack-poisoned]
